### PR TITLE
OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by tafaramafemba**
### In this pull request I managed to: 

1. Create a class Nameable.
2. Implement a method called correct_name that will raise a NotImplementedError.
3. Turn the Person class to Nameable

- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method correct_name implemented. It should simply return the name attribute.

4. Prepare base Decorator

- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the correct_name method that returns the result of the correct_name method of the @nameable.

5. Prepare CapitalizeDecorator and TrimmerDecorator

For the CapitalizeDecorator:

- Create a class that inherits from the base Decorator class.
- Implement a method correct_name that capitalizes the output of @nameable.correct_name.

For the TrimmerDecorator:

- Create a class that inherits from the base Decorator class.
- Implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name(name)
+    name.downcase.capitalize[0, 10].strip
+  end
+end
```
Comment (medaminedev66) [inline nameable.rb]: - [x] The method 'correct_name' in the Naneable class should be rasing an error.

File: person.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require './nameable'
+
 class Person
   attr_reader :id
```
Comment (medaminedev66) [inline person.rb]: - [x] The `Person` class should inherit from `Nameable` class. It also should have a method `correct_name` that returns the name attribute.

File: nameable.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class Nameable
```
Comment (medaminedev66) [inline nameable.rb:1]: - [x] Following the project requirements, you should create a Decorator class that it inherits from Nameable.
Decorator class:
- In the constructor assign a nameable object from params to an instance variable.
- Implement the correct_name method that returns the result of the correct_name method of the @nameable.

Comment (medaminedev66) [inline nameable.rb:1]: - [x] Following the project requirements, you should create CapitalizeDecorator class that inherits from the base Decorator class.
 - In the CapitalizeDecorator class, implement a method correct_name that capitalizes the output of @nameable.correct_name.

Comment (medaminedev66) [inline nameable.rb:1]: - [x] Following the project requirements, you should create TrimmerDecorator class that inherits from the base Decorator class:
- In TrimmerDecorator class, implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

**Review by medaminedev66 — CHANGES_REQUESTED** [review]
Hi @tafaramafemba  😃, :wave:, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- There are no linter errors ✔️
- You have a descriptive PR ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: nameable.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+  def correct_name
+    raise NotImplementedError
```
Comment (FrederickMih) [inline nameable.rb]: - [Optional] Let's modify the error raising like this:

  `raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"`

File: base_decorator.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+require './nameable'
+
+class Decorator < Nameable
+  def initialize
+    super()
+    @nameable = Nameable.new
+  end
```
Comment (FrederickMih) [inline base_decorator.rb]: - [ ] You would need to pass a nameable object as a parameter in the constructor
- [x] You are required to assign the nameable object to an instance variable. Hints:

```ruby
        class Decorator < Nameable
           attr_accessor :nameable
  
            def initialize(nameable)
               @nameable = nameable
            end
          '
          '
          '
```

File: base_decorator.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+  def correct_name
+    @nameable.correct_name(@name)
```
Comment (FrederickMih) [inline base_decorator.rb]: - [x] You don't need to pass the @name instance variable. Let's go like this: `@nameable.correct_name`

**Review by FrederickMih — CHANGES_REQUESTED** [review]
Hi @tafaramafemba , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️

- [Optional] Let's make our README file more descriptive. Provide more information on the `Getting Started` section so that someone without enough skill in programming can easily set up the project locally. The links in the `Author` and `License` sections are not clickable. Let's make them clickable. By fixing this, our README file would become professional.

_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by gemmen29 — APPROVED** [review]
Hi @tafaramafemba, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**